File: combo_engine.py
"""
Project Elements - Combo Engine (Phase 2)

無属性ジョーカーの自動最適化と属性コンボ判定、ダメージ計算を担う。
"""
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum
from itertools import product
from typing import List, Tuple

from models import Attribute, Card, Character


# 基本属性のリスト（TYPELESS を除く）
BASIC_ATTRIBUTES = [Attribute.FIRE, Attribute.WATER, Attribute.LIGHT]


class ComboType(Enum):
    FLASH = "flash"       # 同色3枚: 2.5x
    RAINBOW = "rainbow"   # 3色: 1.5x
    PAIR = "pair"         # 同色2枚 + 異色1枚: ペア部分1.2x / 残り1.0x
    NONE = "none"         # 役なし: 1.0x


@dataclass
class ComboResult:
    combo_type: ComboType
    resolved_attributes: List[Attribute]

    def get_multipliers(self) -> List[float]:
        """
        各カード（インデックス順）に対する倍率リストを返す。
        FLASH/RAINBOW は全カード均等倍率、PAIR はペア2枚に1.2x、残り1枚に1.0x。
        """
        if self.combo_type == ComboType.FLASH:
            return [2.5, 2.5, 2.5]
        if self.combo_type == ComboType.RAINBOW:
            return [1.5, 1.5, 1.5]
        if self.combo_type == ComboType.PAIR:
            from collections import Counter
            counts = Counter(self.resolved_attributes)
            pair_attr = max(counts, key=lambda a: counts[a])
            return [1.2 if a == pair_attr else 1.0 for a in self.resolved_attributes]
        return [1.0, 1.0, 1.0]

    @property
    def total_multiplier(self) -> float:
        """コンボ全体の「代表倍率」（FLASH=2.5, RAINBOW=1.5, PAIR=1.2, NONE=1.0）。
        優先度比較に使用する。"""
        priority = {
            ComboType.FLASH: 2.5,
            ComboType.RAINBOW: 1.5,
            ComboType.PAIR: 1.2,
            ComboType.NONE: 1.0,
        }
        return priority[self.combo_type]


def _judge_combo(attrs: List[Attribute]) -> ComboResult:
    """
    3枚の属性リスト（TYPELESS なし）からコンボ役を判定する。
    """
    if len(attrs) != 3:
        raise ValueError(f"カードは3枚必要です（受け取った枚数: {len(attrs)}）")

    unique = set(attrs)

    # フラッシュ: 3枚すべて同属性
    if len(unique) == 1:
        return ComboResult(ComboType.FLASH, attrs)

    # レインボー: 3色すべて揃う
    if unique == set(BASIC_ATTRIBUTES):
        return ComboResult(ComboType.RAINBOW, attrs)

    # ペア: 2種類（2枚 + 1枚）
    if len(unique) == 2:
        return ComboResult(ComboType.PAIR, attrs)

    return ComboResult(ComboType.NONE, attrs)


def resolve_jokers(
    cards: List[Card],
    leader_attribute: Attribute = Attribute.FIRE,
) -> ComboResult:
    """
    手札3枚の TYPELESS（ジョーカー）を最適な基本属性に自動変換してコンボ結果を返す。

    変換の優先順位: FLASH (2.5x) > RAINBOW (1.5x) > PAIR (1.2x) > NONE (1.0x)

    - TYPELESS が 0 枚: そのまま判定。
    - TYPELESS が 1〜2 枚: 全置換パターンを試し最高倍率を選ぶ。
    - TYPELESS が 3 枚: リーダー属性のフラッシュとして扱う。
    """
    if len(cards) != 3:
        raise ValueError(f"カードは3枚必要です（受け取った枚数: {len(cards)}）")

    typeless_indices = [i for i, c in enumerate(cards) if c.attribute == Attribute.TYPELESS]

    # ジョーカーなし
    if not typeless_indices:
        attrs = [c.attribute for c in cards]
        return _judge_combo(attrs)

    # 全3枚がジョーカー
    if len(typeless_indices) == 3:
        resolved = [leader_attribute] * 3
        return ComboResult(ComboType.FLASH, resolved)

    # 1〜2枚がジョーカー: 全置換パターンを列挙して最高役を選ぶ
    base_attrs = [c.attribute for c in cards]
    best: ComboResult | None = None

    replacement_combinations = product(BASIC_ATTRIBUTES, repeat=len(typeless_indices))
    for replacements in replacement_combinations:
        trial = base_attrs[:]
        for idx, replacement in zip(typeless_indices, replacements):
            trial[idx] = replacement
        result = _judge_combo(trial)
        if best is None or result.total_multiplier > best.total_multiplier:
            best = result

    return best  # type: ignore[return-value]  # typeless が 1〜2 枚なら必ず候補がある


@dataclass
class DamageResult:
    character_damages: List[float]
    combo_result: ComboResult

    @property
    def total_damage(self) -> float:
        return sum(self.character_damages)


def calculate_damage(
    cards: List[Card], # プレイされた3枚のカード
    characters: List[Character], # パーティメンバー全員（3体）
    leader_attribute: Attribute = Attribute.FIRE,
) -> DamageResult:
    """
    場にプレイされたカードと、パーティキャラクター全員のステータスから各ダメージを計算する。

    - 属性コンボ倍率（ジョーカー変換済み）を算出し、全キャラクターの基礎攻撃力に乗算する。
    - 所有者一致ボーナス（1.2x）:
      プレイされた3枚のカードの中に、自身の所有するカードが1枚でも含まれているキャラクターの攻撃に適用する。
    - 生存している全キャラクターが攻撃を行う（自身のカードが選ばれなかったキャラもボーナスなしで攻撃する）。

    Args:
        cards: 場にプレイされた3枚のカード
        characters: パーティメンバー全員のリスト（カードとのインデックス同期は不要）
        leader_attribute: 全ジョーカーの場合のフォールバック属性

    Returns:
        各キャラクターが与えるダメージリストとコンボ結果を含む DamageResult
    """
    if len(cards) != 3:
        raise ValueError("プレイされるカードは3枚である必要があります")

    if len(characters) == 0:
        raise ValueError("攻撃を行うキャラクターがパーティに存在しません")

    # 1. コンボ判定と全体倍率の取得
    combo_result = resolve_jokers(cards, leader_attribute)
    combo_multiplier = combo_result.total_multiplier

    # 2. 場に出たカードの所有者（オーナー）をリスト化
    played_owners = [card.owner for card in cards]

    character_damages: List[float] = []

    # 3. 全キャラクターがそれぞれ攻撃を行う
    for char in characters:
        if not char.is_alive:
            character_damages.append(0.0)
            continue

        # 基礎攻撃力 × コンボ全体倍率
        damage = char.attack_power * combo_multiplier

        # 4. 所有者一致ボーナスの判定
        # 場に出た3枚の中に、自分のカードが1枚でもあればボーナス
        if char.name in played_owners:
            damage *= 1.2

        character_damages.append(damage)

    return DamageResult(character_damages=character_damages, combo_result=combo_result)

File: Ver-C#/Scripts/Logic/ComboEngine.cs
// Ver-C#/Scripts/Logic/ComboEngine.cs
// combo_engine.py の移植。
// ジョーカー（無属性）の自動最適化、コンボ役判定、最終ダメージ計算を担う。

using System;
using System.Collections.Generic;
using System.Linq;
using ProjectElements.Core;
using ProjectElements.Data;
using ProjectElements.Models;

namespace ProjectElements.Logic
{
    // =========================================================
    // ComboResult: コンボ判定の結果を保持するデータクラス
    // =========================================================

    /// <summary>
    /// ジョーカー変換後の属性リストと成立したコンボ種別を保持する。
    /// Python の ComboResult dataclass に相当。
    /// </summary>
    public class ComboResult
    {
        /// <summary>成立したコンボ役。</summary>
        public ComboType ComboType { get; }

        /// <summary>
        /// ジョーカー変換済みの属性配列（要素数=3、Typeless なし）。
        /// インデックスはプレイされたカードの順序と対応している。
        /// </summary>
        public Attribute[] ResolvedAttributes { get; }

        public ComboResult(ComboType comboType, Attribute[] resolvedAttributes)
        {
            ComboType          = comboType;
            ResolvedAttributes = resolvedAttributes;
        }

        /// <summary>
        /// 各カード（インデックス順）に対するダメージ倍率を返す。
        /// Flash/Rainbow は全カード均等、Pair はペア2枚に1.2x・残り1枚に1.0x。
        /// </summary>
        public float[] GetMultipliers()
        {
            switch (ComboType)
            {
                case ComboType.Flash:
                    return new[] { 2.5f, 2.5f, 2.5f };

                case ComboType.Rainbow:
                    return new[] { 1.5f, 1.5f, 1.5f };

                case ComboType.Pair:
                    // 2回登場する属性がペアの属性
                    Attribute pairAttr = ResolvedAttributes
                        .GroupBy(a => a)
                        .OrderByDescending(g => g.Count())
                        .First().Key;
                    return ResolvedAttributes
                        .Select(a => a == pairAttr ? 1.2f : 1.0f)
                        .ToArray();

                default:
                    return new[] { 1.0f, 1.0f, 1.0f };
            }
        }

        /// <summary>
        /// コンボ種別の代表倍率。ジョーカー最適化の優先度比較に使用する。
        /// Flash=2.5 > Rainbow=1.5 > Pair=1.2 > None=1.0
        /// </summary>
        public float TotalMultiplier
        {
            get
            {
                switch (ComboType)
                {
                    case ComboType.Flash:   return 2.5f;
                    case ComboType.Rainbow: return 1.5f;
                    case ComboType.Pair:    return 1.2f;
                    default:                return 1.0f;
                }
            }
        }
    }

    // =========================================================
    // DamageResult: ダメージ計算結果を保持するデータクラス
    // =========================================================

    /// <summary>
    /// 各キャラクターの最終ダメージとコンボ結果をまとめて保持する。
    /// Python の DamageResult dataclass に相当。
    /// </summary>
    public class DamageResult
    {
        /// <summary>
        /// キャラクターごとの最終ダメージ。
        /// インデックスは CalculateDamage に渡した characters 配列と対応している。
        /// </summary>
        public float[] CharacterDamages { get; }

        /// <summary>今ターンのコンボ判定結果。</summary>
        public ComboResult ComboResult { get; }

        /// <summary>全キャラクターのダメージ合計。</summary>
        public float TotalDamage => CharacterDamages.Sum();

        public DamageResult(float[] characterDamages, ComboResult comboResult)
        {
            CharacterDamages = characterDamages;
            ComboResult      = comboResult;
        }
    }

    // =========================================================
    // ComboEngine: コンボ処理のメインロジック（静的ユーティリティ）
    // =========================================================

    /// <summary>
    /// 無属性ジョーカーの最適化、コンボ役判定、ダメージ計算をまとめた静的クラス。
    /// Python の combo_engine.py 全体に相当。
    /// </summary>
    public static class ComboEngine
    {
        // 基本属性リスト（Typeless を除く）
        // Python の BASIC_ATTRIBUTES に相当
        private static readonly Attribute[] BasicAttributes =
            { Attribute.Fire, Attribute.Water, Attribute.Light };

        // ---------------------------------------------------------
        // ① コンボ役の純粋判定（Typeless なし前提）
        // ---------------------------------------------------------

        /// <summary>
        /// Typeless を含まない3属性の配列からコンボ役を判定して返す。
        /// Python の _judge_combo() に相当。
        /// </summary>
        private static ComboResult JudgeCombo(Attribute[] attrs)
        {
            var unique = new HashSet<Attribute>(attrs);

            // フラッシュ: 3枚すべて同属性
            if (unique.Count == 1)
                return new ComboResult(ComboType.Flash, attrs);

            // レインボー: 3種類すべて異なる（Typeless 除去済みなので Fire/Water/Light の3色確定）
            if (unique.Count == 3)
                return new ComboResult(ComboType.Rainbow, attrs);

            // ペア: 2種類（2枚 + 1枚）
            if (unique.Count == 2)
                return new ComboResult(ComboType.Pair, attrs);

            return new ComboResult(ComboType.None, attrs);
        }

        // ---------------------------------------------------------
        // ② ジョーカー全置換パターンの列挙
        //    Python の itertools.product(BASIC_ATTRIBUTES, repeat=n) に相当
        // ---------------------------------------------------------

        /// <summary>
        /// count 個の置換スロットに対して、基本属性の全組み合わせを列挙する。
        /// count=1 → 3通り、count=2 → 9通り（3×3）。
        /// </summary>
        private static IEnumerable<Attribute[]> GetReplacementCombinations(int count)
        {
            if (count == 1)
            {
                // 1ジョーカー: Fire / Water / Light の3通りを試す
                foreach (var a in BasicAttributes)
                    yield return new[] { a };
            }
            else // count == 2
            {
                // 2ジョーカー: 3×3 = 9通りの直積を試す
                foreach (var a in BasicAttributes)
                    foreach (var b in BasicAttributes)
                        yield return new[] { a, b };
            }
        }

        // ---------------------------------------------------------
        // ③ ジョーカー自動最適化（メイン公開メソッド）
        // ---------------------------------------------------------

        /// <summary>
        /// 手札3枚の Typeless（ジョーカー）を最も高倍率になる基本属性に自動変換し、
        /// コンボ判定結果を返す。
        ///
        /// 変換優先度: Flash(2.5x) > Rainbow(1.5x) > Pair(1.2x) > None(1.0x)
        ///
        /// ・Typeless 0枚 → そのまま判定
        /// ・Typeless 1〜2枚 → 全置換パターンを試して最高役を選択
        /// ・Typeless 3枚 → リーダー属性のフラッシュとして扱う
        ///
        /// Python の resolve_jokers() に相当。
        /// </summary>
        public static ComboResult ResolveJokers(
            CardData[]  cards,
            Attribute   leaderAttribute = Attribute.Fire)
        {
            if (cards.Length != 3)
                throw new ArgumentException(
                    $"カードは3枚必要です（受け取った枚数: {cards.Length}）");

            // Typeless のインデックスを収集
            var typelessIndices = cards
                .Select((card, idx) => (card, idx))
                .Where(t => t.card.attribute == Attribute.Typeless)
                .Select(t => t.idx)
                .ToList();

            // ジョーカーなし: そのまま判定
            if (typelessIndices.Count == 0)
                return JudgeCombo(cards.Select(c => c.attribute).ToArray());

            // 全3枚ジョーカー: リーダー属性のフラッシュ
            if (typelessIndices.Count == 3)
                return new ComboResult(
                    ComboType.Flash,
                    new[] { leaderAttribute, leaderAttribute, leaderAttribute });

            // 1〜2枚ジョーカー: 全置換パターンを試して最高役を選ぶ
            Attribute[] baseAttrs = cards.Select(c => c.attribute).ToArray();
            ComboResult best      = null;

            foreach (var replacement in GetReplacementCombinations(typelessIndices.Count))
            {
                // 元の配列を壊さないようにコピーしてから置換
                Attribute[] trial = (Attribute[])baseAttrs.Clone();
                for (int i = 0; i < typelessIndices.Count; i++)
                    trial[typelessIndices[i]] = replacement[i];

                ComboResult candidate = JudgeCombo(trial);
                if (best == null || candidate.TotalMultiplier > best.TotalMultiplier)
                    best = candidate;
            }

            return best;
        }

        // ---------------------------------------------------------
        // ④ 最終ダメージ計算
        // ---------------------------------------------------------

        /// <summary>
        /// プレイされた3枚のカードとパーティメンバーから各キャラクターの最終ダメージを計算する。
        ///
        /// 計算式:
        ///   基礎攻撃力 × コンボ代表倍率 [× 所有者一致ボーナス 1.2x]
        ///
        /// ・生存キャラクター全員が攻撃する（自分のカードが場に出なかった場合もボーナスなしで攻撃）。
        /// ・所有者一致ボーナス: プレイされた3枚の中に自分の所持カードが1枚でも含まれていれば適用。
        ///   ScriptableObject の参照同一性で判定する。
        ///
        /// Python の calculate_damage() に相当。
        /// </summary>
        public static DamageResult CalculateDamage(
            CardData[]       playedCards,
            BattleCharacter[] characters,
            Attribute        leaderAttribute = Attribute.Fire)
        {
            if (playedCards.Length != 3)
                throw new ArgumentException("プレイされるカードは3枚である必要があります");
            if (characters.Length == 0)
                throw new ArgumentException("攻撃を行うキャラクターがパーティに存在しません");

            // 1. コンボ判定（ジョーカー変換含む）
            ComboResult comboResult     = ResolveJokers(playedCards, leaderAttribute);
            float       comboMultiplier = comboResult.TotalMultiplier;

            var characterDamages = new float[characters.Length];

            for (int i = 0; i < characters.Length; i++)
            {
                BattleCharacter chara = characters[i];

                // 戦闘不能キャラはダメージ 0
                if (!chara.IsAlive)
                {
                    characterDamages[i] = 0f;
                    continue;
                }

                // 2. 基礎攻撃力 × コンボ倍率
                float damage = chara.AttackPower * comboMultiplier;

                // 3. 所有者一致ボーナス判定
                //    ScriptableObject はアセットの参照同一性（== 演算子）で比較する
                bool hasOwnerBonus = playedCards.Any(played =>
                    chara.Cards.Any(owned => owned == played));

                if (hasOwnerBonus)
                    damage *= 1.2f;

                characterDamages[i] = damage;
            }

            return new DamageResult(characterDamages, comboResult);
        }
    }
}
